File: 456-132-pattern/456-132-pattern.py
class Solution(object):
    def find132pattern(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        l = len(nums)
        
        if l < 3:
            return False
        
        mins = [0]*l
        mins[0] = nums[0]
        for i in range(1, l):
            mins[i] = min(mins[i-1], nums[i])
        stack = []
        for i in range(l-1, -1, -1):
            if stack and nums[i] > stack[-1]:
                while stack and nums[i] > stack[-1]:
                    n = stack.pop()
                    if n > mins[i]:
                        return True
            stack.append(nums[i])
        return False

File: 456-132-pattern/456-132-pattern.java
class Solution {
    public boolean find132pattern(int[] nums) {
        if (nums.length < 3) {
            return false;
        }
        int[] left_min = new int[nums.length];
        Stack <Integer> stack = new Stack < > ();
        left_min[0] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            left_min[i] = Math.min(left_min[i-1], nums[i]);
        }
        for (int i = nums.length - 1; i >= 0; i--) {
            if (stack.size() > 0 && nums[i] > stack.peek()) {
                while (stack.size() > 0 && nums[i] > stack.peek()) {
                    int n = stack.pop();
                    if (n > left_min[i]) {
                        return true;
                    }
                }
            }
            stack.push(nums[i]);
        }
        return false;
    }
}